scrapy/extensions/batches.py

```diff
@@ -0,0 +1,132 @@
+"""
+Extensions for batch processing and support.
+"""
+
+import re
+import time
+from typing import Any, BinaryIO, Dict
+
+
+class BatchHandler:
+    """
+    The default batch handler for handling batch formation.
+
+    To activate batching using the default batch handler, assign any criteria in the :ref:`feeds-options
+    <feed-options>` of a feed with the desired limit. When using more than one criteria,
+    whichever criteria is exceeded first will trigger the formation of the new batch.
+
+    Accepted criteria:
+
+    - ``batch_item_count``: Assign a limit on the number of items a batch can have.
+
+        The limit must be an integer value.
+
+    - ``batch_file_size``: Assign a soft limit to the size of a batch file.
+
+        The limit must be a string in the format <``SIZE``><``STORAGE-UNIT``>, where ``STORAGE-UNIT``
+        must be byte unit based on powers of 2(KiB, MiB, GiB, TiB) or powers of 10(kB, MB, GB, TB).
+        Eg: 200MB, 100KiB.
+
+    - ``batch_duration``: Assign a soft limit on the duration a batch stays active.
+
+        Duration must be a string in the format ``hours:minutes:seconds``. Eg: 1:0:0 for 1 hour,
+        0:30:0 for a 30 minute duration.
+
+    Criteria can also be set in settings which will be assigned to all the feeds. But
+    criteria assigned through feed-options will have higher priority.
+
+    Equivalent settings criteria:
+
+    .. setting:: FEED_EXPORT_BATCH_ITEM_COUNT
+
+    - ``FEED_EXPORT_BATCH_ITEM_COUNT``: Equivalent to ``batch_item_count``.
+
+        Default: ``0``
+
+    .. setting:: FEED_EXPORT_BATCH_FILE_SIZE
+
+    - ``FEED_EXPORT_BATCH_FILE_SIZE``: Equivalent to ``batch_file_size``.
+
+        Default: ``0B``
+
+    .. setting:: FEED_EXPORT_BATCH_DURATION
+
+    - ``FEED_EXPORT_BATCH_DURATION``: Equivalent to ``batch_duration``.
+
+        Default: ``0:0:0``
+    """
+
+    def __init__(self, feed_options: Dict[str, Any]) -> None:
+        # get limits from feed_settings
+        self.max_item_count: int = feed_options["batch_item_count"]
+        self.max_seconds: float = self._in_seconds(feed_options["batch_duration"])
+        self.max_file_size: int = self._in_bytes(feed_options["batch_file_size"])
+        # initialize batch state attributes
+        self.item_count: int = 0
+        self.elapsed_seconds: float = 0
+        self.file_size: int = 0
+        self.batch_id: int = 0
+        # misc attributes
+        self.file: BinaryIO
+        self.start_time: float
+        self.enabled: bool = any([self.max_item_count, self.max_seconds, self.max_file_size])
+
+    def item_added(self) -> None:
+        """
+        Update batch state attributes.
+        """
```

Comment:
For a class implementing an interface, such as this one, you usually do not need to document the interface methods. The documentation of a method is meant to describe what it does, not how it does it, and that’s usually already covered in the documentation of the interface.

If your docstrings are meant to describe how a method works, and you feel this information is needed as a comment because the implementation or some aspect of it is not clear from code, you should use code comments for such information, not a docstring.

Reply:
Yeah sorry I repeated this in postprocessing as well.